### PR TITLE
[feat]: [CDS-50311]: Added Logic to handle finalise and next step execution in TaskChainExecutables

### PR DESCRIPTION
**Pull request by arvind-choudhary-h**
## Describe your changes
1. CdTaskChainExecutable is modified to support saving Node Info provided by the Delegate response in both intermediate and final step.
2. `stage-instance-info` API is moved to the internal

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49839)

### CODE REVIEW

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 11 of 11 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49839)* status: 11 file(s) to review. 0 of 2 approvals obtained.

**Comment by arvind-choudhary-h** [comment]
trigger uts

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/cdng/execution/StepExecutionInstanceInfo.java [hunk 1]
```diff
@@ -16,7 +16,7 @@
 
 @Data
 @Builder
-@RecasterAlias("io.harness.cdng.execution.StepExecutionInstanceInfo")
```
Comment (abhinav-harness) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/cdng/execution/StepExecutionInstanceInfo.java:19]: why do you want recaster alias to change?
Wont it cause bg compatibility issues ?

Comment (arvind-choudhary-h) [reply 950-delegate-tasks-beans/src/main/java/io/harness/delegate/cdng/execution/StepExecutionInstanceInfo.java:19]: No no, it's not being used yet.

**Comment by yogesh-chauhan** [comment]
trigger uts

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 1 of 1 files at r2, 7 of 7 files at r3, 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49839)* status: Some of the checks are too old. 23 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by namantalaycha10 — APPROVED** [review]

**Comment by yogesh-chauhan** [comment]
trigger uts_JDK11

**Comment by namantalaycha10** [comment]
trigger review

**Comment by arvind-choudhary-h** [comment]
trigger ut6

**Review by mohitg0795 — APPROVED** [review]
> Reviewed 9 of 11 files at r1, 1 of 1 files at r2, 7 of 7 files at r3, 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49839)* status: 1. unresolved discussions.

**Review by harshjain12 — APPROVED** [review]

**Review by ankittiwari-harness — APPROVED** [review]
approved pipeline svc alias.txt change

**Review by archit-harness — APPROVED** [review]
> Reviewed 9 of 11 files at r1, 1 of 1 files at r2, 7 of 7 files at r3, 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49839)* status: 1. unresolved discussions.

**Pull request merged**